smolyay/surrogate.py

```diff
@@ -1,546 +1,720 @@
+            )
+
+        # create lookup table and solve for all the basis functions
+        lookup_table = []
+        lookup_table_derivative = []
+        for dim in range(self.num_dimensions):
+            lookup_table.append(self.basis_sets[dim](X[:, dim], self.domain[dim]))
+            lookup_table_derivative.append(
+                self.basis_sets[dim].derivative(X[:, dim], domain=self.domain[dim])
+            )
+
+        # create basis matrix
+        basis_matrix = numpy.zeros(
+            (len(X) * self.num_dimensions, len(self._index_combinations)),
+            dtype=(
+                complex
+                if any(
+                    any(bf._is_complex for bf in basis_set)
+                    for basis_set in self.basis_sets
+                )
+                else float
+            ),
+        )
+
+        # use lookup table to solve for each term
+        for d in range(self.num_dimensions):
+            for term, ic in enumerate(self._index_combinations):
+                if self.num_dimensions > 1:
+                    basis_matrix[d :: self.num_dimensions, term] = numpy.prod(
+                        [
+                            (
+                                lookup_table_derivative[dim][ic[dim], :]
+                                if dim == d
+                                else lookup_table[dim][ic[dim], :]
+                            )
+                            for dim in range(len(ic))
+                        ],
+                        axis=0,
+                    )
+                else:
+                    basis_matrix[d :: self.num_dimensions, term] = (
+                        lookup_table_derivative[d][ic[d], :]
+                    )
+
+        data = numpy.reshape(y, (self.num_dimensions * len(X),))
+
+        # solve for coefficients
+        if self.regularization is None or numpy.iscomplexobj(basis_matrix):
+            self._coefficients = numpy.linalg.lstsq(basis_matrix, data, rcond=None)[0]
+        elif isinstance(self.regularization, L2Regularization):
+            regressor = sklearn.linear_model.Ridge(
+                alpha=self.regularization.alpha, fit_intercept=False
+            )
+            self._coefficients = numpy.squeeze(regressor.fit(basis_matrix, data).coef_)
+        elif isinstance(self.regularization, L1Regularization):
+            regressor = sklearn.linear_model.Lasso(
+                alpha=self.regularization.alpha, fit_intercept=False
+            )
+            self._coefficients = numpy.squeeze(regressor.fit(basis_matrix, data).coef_)
         else:
-            for i in range(new.shape[0]):
-                numpy.clip(new_x[...,i], new[i,0], new[i,1],out=new_x[...,i])
-        return new_x
+            self._coefficients = numpy.linalg.lstsq(basis_matrix, data, rcond=None)[0]
+        self._needs_fit = False
+        self._fit_gradient_flag = True
+
+        # determine integration constant if possible
+        if not y0 is None:
```

Comment:
Try to get in the habit of using this syntax. `is not` is an operator in Python.
```suggestion
        if y0 is not None:
```